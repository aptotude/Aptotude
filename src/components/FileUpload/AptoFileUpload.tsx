import * as React from 'react';
import Dropzone from 'react-dropzone';
import classNames from 'classnames';
import './aptoFileUpload.scss';
import { AptoParagraph } from '../Paragraph';
import { AptoIcon } from '../Icon';
import { AptoList } from '../List';
import { AptoButton } from '../Button';
import { AptoListItem } from '../ListItem';
import { UploadIcon } from './UploadIcon';

interface AptoFileUploadProps {
  name: string;
  value?: string | undefined | null;
  onDrop?: (acceptedFiles: any, rejectedFiles: any) => void;
  onCancel?: (originalValue: string | null) => void;
  onRemoveImage?: () => void;
  maxSize?: number;
  minSize?: number;
  multiple?: boolean;
  accept?: string | undefined;
  imageComponent?:
    | React.ComponentClass<any>
    | React.StatelessComponent<any>
    | any;
}

interface AptoFileUploadState {
  files: any;
  hasPreviousImage: boolean;
}

export class AptoFileUpload extends React.Component<
  AptoFileUploadProps,
  AptoFileUploadState
> {
  public static defaultProps = {
    maxSize: 5242880, // 5mb
    minSize: 0,
    accept: 'image/png,image/jpeg',
    value: null,
    multiple: false
  };

  public originalValue: string | null = null;

  constructor(props: AptoFileUploadProps) {
    super(props);
    this.originalValue = this.props.value || null;
    this.state = {
      files: [],
      hasPreviousImage:
        this.originalValue && this.originalValue !== '' ? true : false
    };
  }

  public removeOldImage = (e: React.MouseEvent) => {
    const { onRemoveImage } = this.props;
    e.preventDefault();

    this.setState({
      hasPreviousImage: false
    });

    if (onRemoveImage) {
      onRemoveImage();
    }
  };

  public onDrop = (acceptedFiles: any, rejectedFiles: any) => {
    const { onDrop } = this.props;

    this.setState({
      files: acceptedFiles
    });

    if (onDrop) {
      onDrop(acceptedFiles, rejectedFiles);
    }
  };

  public handleCancel = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    this.setState({
      hasPreviousImage: this.originalValue ? true : false
    });

    this.onCancel();
  };

  public onCancel = () => {
    const { onCancel } = this.props;

    this.setState({
      files: []
    });

    if (onCancel) {
      onCancel(this.originalValue);
    }
  };

  public render() {
    if (this.state.hasPreviousImage) {
      const { imageComponent: Component } = this.props;
      let image = null;
      console.log(typeof Component);
      if (this.originalValue && typeof Component === 'object') {
        image = Component;
      } else if (this.originalValue && Component) {
        image = <Component />;
      } else if (this.originalValue) {
        image = <img src={this.originalValue} alt="" />;
      }
      return (
        <div className="dropzone-previousImage">
          {image}
          <br />
          <AptoButton
            className="dropzone-change"
            kind="link"
            onClick={this.removeOldImage}
          >
            Change
          </AptoButton>
        </div>
      );
    }

    const { files } = this.state;
    const acceptedList =
      files && files.length
        ? files.map((f: any) => {
            return (
              <AptoListItem key={f.name || 'unknown'}>
                {f.name || 'unknown'}
              </AptoListItem>
            );
          })
        : null;
    return (
      <Dropzone
        maxSize={this.props.maxSize}
        minSize={this.props.minSize}
        multiple={this.props.multiple}
        name={this.props.name}
        onFileDialogCancel={this.onCancel}
        accept={this.props.accept}
        onDrop={this.onDrop}
      >
        {({ getRootProps, getInputProps, isDragActive }) => {
          return (
            <div
              {...getRootProps()}
              className={classNames('dropzone', {
                'dropzone--isActive': isDragActive
              })}
            >
              <AptoIcon
                size="5"
                icon={UploadIcon}
                className="dropzone-addIcon"
              />
              <input {...getInputProps()} />
              {isDragActive ? (
                <AptoParagraph>Drop the file here</AptoParagraph>
              ) : (
                <div>
                  <span className="AptoButton AptoButton--button AptoButton--primary">
                    Choose File{this.props.multiple ? 's' : null} To Upload
                  </span>
                  <AptoParagraph>
                    or drag and drop file{this.props.multiple ? 's' : null} here
                  </AptoParagraph>
                </div>
              )}
              {acceptedList && (
                <AptoList type="ul" className="fileList">
                  {acceptedList}
                </AptoList>
              )}
              {this.originalValue && (
                <AptoButton
                  kind="link"
                  className="dropzone-cancel"
                  onClick={this.handleCancel}
                >
                  Cancel
                </AptoButton>
              )}
            </div>
          );
        }}
      </Dropzone>
    );
  }
}
