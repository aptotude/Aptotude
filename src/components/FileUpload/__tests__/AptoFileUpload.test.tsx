import * as React from 'react';
import { render, cleanup, fireEvent, wait } from 'react-testing-library';
import { AptoFileUpload } from '../AptoFileUpload';

afterEach(() => {
  cleanup();
});

const Img = () => {
  return <div>some custom image thingy</div>;
};

describe('AptoFileUpload', () => {
  it('Renders Component', () => {
    const { container } = render(<AptoFileUpload name="superName" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders Custom Image Component', () => {
    const { container } = render(
      <AptoFileUpload value="foo.jpg" imageComponent={Img} name="superName" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders Component with previous image', () => {
    const { container } = render(
      <AptoFileUpload value="http://foo.com/j.jpg" name="superName" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('toggles change', async () => {
    let hitRemove = false;
    const { container } = render(
      <AptoFileUpload
        onRemoveImage={() => {
          hitRemove = true;
        }}
        value="http://foo.com/j.jpg"
        name="superName"
      />
    );
    const btn = container.querySelector('.dropzone-change');
    if (btn) {
      fireEvent.click(btn);
    }
    await wait(() => {
      const el = container.querySelector('.dropzone');
      expect(el).not.toBeUndefined();
      expect(hitRemove).toBeTruthy();
    });
  });

  it('toggles cancel', async () => {
    let hitCancel = false;
    let old: string | null = null;
    const { container } = render(
      <AptoFileUpload
        onCancel={oldImage => {
          hitCancel = true;
          old = oldImage;
        }}
        value="http://foo.com/j.jpg"
        name="superName"
      />
    );
    const changeBtn = container.querySelector('.dropzone-change');
    if (changeBtn) {
      fireEvent.click(changeBtn);
    }
    await wait(async () => {
      const cancelBtn = container.querySelector('.dropzone-cancel');
      if (cancelBtn) {
        fireEvent.click(cancelBtn);
      }
    });

    await wait(() => {
      const el = container.querySelector('.dropzone-previousImage');
      expect(el).not.toBeUndefined();
      expect(hitCancel).toBeTruthy();
      expect(old).toEqual('http://foo.com/j.jpg');
    });
  });
});
