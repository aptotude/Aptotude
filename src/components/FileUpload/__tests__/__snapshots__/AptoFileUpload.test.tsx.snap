// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AptoFileUpload Renders Component 1`] = `
<div
  class="dropzone"
  tabindex="0"
>
  <span
    class="AptoIcon AptoIcon--size5 dropzone-addIcon"
  >
    <svg
      focusable="false"
      height="100%"
      preserveAspectRatio="xMidYMid meet"
      viewBox="0 0 24 24"
      width="100%"
    >
      <defs>
        <path
          d="M24 24H0V0h24v24z"
          id="addPhoto_svg__a"
        />
      </defs>
      <clippath
        id="addPhoto_svg__b"
      >
        <use
          overflow="visible"
          xlink:href="#addPhoto_svg__a"
        />
      </clippath>
      <path
        clip-path="url(#addPhoto_svg__b)"
        d="M3 4V1h2v3h3v2H5v3H3V6H0V4h3zm3 6V7h3V4h7l1.83 2H21c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V10h3zm7 9c2.76 0 5-2.24 5-5s-2.24-5-5-5-5 2.24-5 5 2.24 5 5 5zm-3.2-5c0 1.77 1.43 3.2 3.2 3.2s3.2-1.43 3.2-3.2-1.43-3.2-3.2-3.2-3.2 1.43-3.2 3.2z"
      />
    </svg>
  </span>
  <input
    accept="image/png,image/jpeg"
    autocomplete="off"
    name="superName"
    style="display: none;"
    tabindex="-1"
    type="file"
  />
  <div>
    <span
      class="AptoButton AptoButton--button AptoButton--primary"
    >
      Choose File
       To Upload
    </span>
    <p
      class="AptoParagraph"
    >
      or drag and drop file
       here
    </p>
  </div>
</div>
`;

exports[`AptoFileUpload Renders Component with previous image 1`] = `
<div
  class="dropzone-previousImage"
>
  <img
    alt=""
    src="http://foo.com/j.jpg"
  />
  <br />
  <button
    class="AptoButton AptoButton--link AptoButton--primary dropzone-change"
  >
    Change
  </button>
</div>
`;

exports[`AptoFileUpload Renders Custom Image Component 1`] = `
<div
  class="dropzone-previousImage"
>
  <div>
    some custom image thingy
  </div>
  <br />
  <button
    class="AptoButton AptoButton--link AptoButton--primary dropzone-change"
  >
    Change
  </button>
</div>
`;
